File: src/memento/control/signals.py
from __future__ import annotations

import hashlib
import json
import sqlite3
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any
from uuid import uuid4


@dataclass(frozen=True, slots=True)
class DreamSignal:
    signal_id: str
    signal_type: str
    entity_refs: tuple[str, ...]
    severity: str
    repo_revision: str
    dedupe_key: str
    status: str
    evidence_hash: str
    evidence_json: str
    first_detected_at: str
    last_detected_at: str
    resolved_revision: str | None


@dataclass(frozen=True, slots=True)
class DetectedSignal:
    signal_type: str
    entity_refs: tuple[str, ...]
    severity: str
    dedupe_key: str
    evidence: dict[str, Any]

    @property
    def evidence_hash(self) -> str:
        return hashlib.sha256(
            json.dumps(self.evidence, sort_keys=True, separators=(",", ":")).encode("utf-8")
        ).hexdigest()


def utcnow() -> str:
    return datetime.now(tz=UTC).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def list_signals(connection: sqlite3.Connection) -> tuple[DreamSignal, ...]:
    rows = connection.execute(
        "SELECT * FROM dream_signals ORDER BY signal_type, dedupe_key"
    ).fetchall()
    return tuple(_row_to_signal(row) for row in rows)


def upsert_detected_signals(
    connection: sqlite3.Connection,
    *,
    repo_revision: str,
    detections: list[DetectedSignal],
) -> tuple[DreamSignal, ...]:
    now = utcnow()
    by_key = {signal.dedupe_key: signal for signal in list_signals(connection)}
    seen_keys = {item.dedupe_key for item in detections}
    with connection:
        for detection in detections:
            evidence_json = json.dumps(detection.evidence, sort_keys=True)
            existing = by_key.get(detection.dedupe_key)
            if existing is None:
                connection.execute(
                    """
                    INSERT INTO dream_signals(
                        signal_id, signal_type, entity_refs_json, severity, repo_revision,
                        dedupe_key, status, evidence_hash, evidence_json,
                        first_detected_at, last_detected_at, resolved_revision
                    ) VALUES(?, ?, ?, ?, ?, ?, 'open', ?, ?, ?, ?, NULL)
                    """,
                    (
                        str(uuid4()),
                        detection.signal_type,
                        json.dumps(list(detection.entity_refs)),
                        detection.severity,
                        repo_revision,
                        detection.dedupe_key,
                        detection.evidence_hash,
                        evidence_json,
                        now,
                        now,
                    ),
                )
                continue
            status = existing.status
            resolved_revision = existing.resolved_revision
            if (
                existing.status in {"resolved", "ignored"}
                and existing.evidence_hash != detection.evidence_hash
            ):
                status = "open"
                resolved_revision = None
            connection.execute(
                """
                UPDATE dream_signals
                SET signal_type = ?, entity_refs_json = ?, severity = ?, repo_revision = ?,
                    status = ?, evidence_hash = ?, evidence_json = ?, last_detected_at = ?,
                    resolved_revision = ?
                WHERE dedupe_key = ?
                """,
                (
                    detection.signal_type,
                    json.dumps(list(detection.entity_refs)),
                    detection.severity,
                    repo_revision,
                    status,
                    detection.evidence_hash,
                    evidence_json,
                    now,
                    resolved_revision,
                    detection.dedupe_key,
                ),
            )
        for existing in by_key.values():
            if existing.dedupe_key in seen_keys:
                continue
            if existing.status in {"open", "acknowledged", "proposed"}:
                connection.execute(
                    "UPDATE dream_signals SET status = 'resolved', resolved_revision = ? WHERE dedupe_key = ?",
                    (repo_revision, existing.dedupe_key),
                )
    return list_signals(connection)


def actionable_signals(connection: sqlite3.Connection) -> tuple[DreamSignal, ...]:
    rows = connection.execute(
        "SELECT * FROM dream_signals WHERE status IN ('open', 'acknowledged') ORDER BY signal_type, dedupe_key"
    ).fetchall()
    return tuple(_row_to_signal(row) for row in rows)


def mark_signals_status(
    connection: sqlite3.Connection, *, dedupe_keys: tuple[str, ...], status: str
) -> None:
    if not dedupe_keys:
        return
    placeholders = ", ".join("?" for _ in dedupe_keys)
    with connection:
        connection.execute(
            f"UPDATE dream_signals SET status = ? WHERE dedupe_key IN ({placeholders})",
            (status, *dedupe_keys),
        )


def set_service_state(connection: sqlite3.Connection, *, key: str, value: str) -> None:
    with connection:
        connection.execute(
            """
            INSERT INTO service_state(key, value, updated_at) VALUES(?, ?, datetime('now'))
            ON CONFLICT(key) DO UPDATE SET value=excluded.value, updated_at=datetime('now')
            """,
            (key, value),
        )


def get_service_state(connection: sqlite3.Connection, *, key: str) -> str | None:
    row = connection.execute("SELECT value FROM service_state WHERE key = ?", (key,)).fetchone()
    return None if row is None else str(row["value"])


def _row_to_signal(row: sqlite3.Row) -> DreamSignal:
    return DreamSignal(
        signal_id=row["signal_id"],
        signal_type=row["signal_type"],
        entity_refs=tuple(str(item) for item in json.loads(row["entity_refs_json"])),
        severity=row["severity"],
        repo_revision=row["repo_revision"],
        dedupe_key=row["dedupe_key"],
        status=row["status"],
        evidence_hash=row["evidence_hash"],
        evidence_json=row["evidence_json"],
        first_detected_at=row["first_detected_at"],
        last_detected_at=row["last_detected_at"],
        resolved_revision=row["resolved_revision"],
    )

File: internal/control/signals.go
package control

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"encoding/json"
	"errors"
	"github.com/google/uuid"
	"github.com/rcarmo/memento/internal/pyjson"
	"strings"
	"time"
)

type DreamSignal struct {
	SignalID                                                                                               string   `json:"signal_id"`
	SignalType                                                                                             string   `json:"signal_type"`
	EntityRefs                                                                                             []string `json:"entity_refs"`
	Severity, RepoRevision, DedupeKey, Status, EvidenceHash, EvidenceJSON, FirstDetectedAt, LastDetectedAt string
	ResolvedRevision                                                                                       *string
}
type DetectedSignal struct {
	SignalType          string
	EntityRefs          []string
	Severity, DedupeKey string
	Evidence            map[string]any
}

func (d DetectedSignal) evidence() (string, string, error) {
	raw, err := pyjson.DumpsCompact(d.Evidence)
	if err != nil {
		return "", "", err
	}
	sum := sha256.Sum256([]byte(raw))
	pretty, _ := pyjson.Dumps(d.Evidence)
	return hex.EncodeToString(sum[:]), pretty, nil
}

type Signals struct {
	DB   *sql.DB
	Now  func() time.Time
	UUID func() string
}

func (s Signals) now() string {
	now := time.Now()
	if s.Now != nil {
		now = s.Now()
	}
	return now.UTC().Truncate(time.Second).Format(time.RFC3339)
}
func (s Signals) id() string {
	if s.UUID != nil {
		return s.UUID()
	}
	return uuid.NewString()
}

const signalColumns = "signal_id,signal_type,entity_refs_json,severity,repo_revision,dedupe_key,status,evidence_hash,evidence_json,first_detected_at,last_detected_at,resolved_revision"

func (s Signals) List(ctx context.Context) ([]DreamSignal, error) {
	return s.list(ctx, "SELECT "+signalColumns+" FROM dream_signals ORDER BY signal_type,dedupe_key")
}
func (s Signals) Actionable(ctx context.Context) ([]DreamSignal, error) {
	return s.list(ctx, "SELECT "+signalColumns+" FROM dream_signals WHERE status IN ('open','acknowledged') ORDER BY signal_type,dedupe_key")
}
func (s Signals) list(ctx context.Context, query string) ([]DreamSignal, error) {
	rows, err := s.DB.QueryContext(ctx, query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	result := []DreamSignal{}
	for rows.Next() {
		item, err := scanSignal(rows)
		if err != nil {
			return nil, err
		}
		result = append(result, item)
	}
	return result, rows.Err()
}
func (s Signals) Upsert(ctx context.Context, revision string, detections []DetectedSignal) ([]DreamSignal, error) {
	existing, err := s.List(ctx)
	if err != nil {
		return nil, err
	}
	byKey := map[string]DreamSignal{}
	for _, item := range existing {
		byKey[item.DedupeKey] = item
	}
	seen := map[string]bool{}
	now := s.now()
	err = WithTransaction(ctx, s.DB, func(tx *sql.Tx) error {
		for _, detection := range detections {
			seen[detection.DedupeKey] = true
			hash, evidence, encodeErr := detection.evidence()
			if encodeErr != nil {
				return encodeErr
			}
			refs, _ := pyjson.Dumps(detection.EntityRefs)
			current, ok := byKey[detection.DedupeKey]
			if !ok {
				if _, e := tx.ExecContext(ctx, "INSERT INTO dream_signals(signal_id,signal_type,entity_refs_json,severity,repo_revision,dedupe_key,status,evidence_hash,evidence_json,first_detected_at,last_detected_at,resolved_revision) VALUES(?,?,?,?,?,?,'open',?,?,?,?,NULL)", s.id(), detection.SignalType, refs, detection.Severity, revision, detection.DedupeKey, hash, evidence, now, now); e != nil {
					return e
				}
				continue
			}
			status, resolved := current.Status, current.ResolvedRevision
			if (status == "resolved" || status == "ignored") && current.EvidenceHash != hash {
				status = "open"
				resolved = nil
			}
			if _, e := tx.ExecContext(ctx, "UPDATE dream_signals SET signal_type=?,entity_refs_json=?,severity=?,repo_revision=?,status=?,evidence_hash=?,evidence_json=?,last_detected_at=?,resolved_revision=? WHERE dedupe_key=?", detection.SignalType, refs, detection.Severity, revision, status, hash, evidence, now, resolved, detection.DedupeKey); e != nil {
				return e
			}
		}
		for _, item := range existing {
			if !seen[item.DedupeKey] && (item.Status == "open" || item.Status == "acknowledged" || item.Status == "proposed") {
				if _, e := tx.ExecContext(ctx, "UPDATE dream_signals SET status='resolved',resolved_revision=? WHERE dedupe_key=?", revision, item.DedupeKey); e != nil {
					return e
				}
			}
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	return s.List(ctx)
}
func (s Signals) MarkStatus(ctx context.Context, keys []string, status string) error {
	if len(keys) == 0 {
		return nil
	}
	marks := strings.TrimSuffix(strings.Repeat("?,", len(keys)), ",")
	args := []any{status}
	for _, key := range keys {
		args = append(args, key)
	}
	_, err := s.DB.ExecContext(ctx, "UPDATE dream_signals SET status=? WHERE dedupe_key IN ("+marks+")", args...)
	return err
}
func SetServiceState(ctx context.Context, db *sql.DB, key, value string) error {
	_, err := db.ExecContext(ctx, "INSERT INTO service_state(key,value,updated_at) VALUES(?,?,datetime('now')) ON CONFLICT(key) DO UPDATE SET value=excluded.value,updated_at=datetime('now')", key, value)
	return err
}
func GetServiceState(ctx context.Context, db *sql.DB, key string) (*string, error) {
	var value string
	err := db.QueryRowContext(ctx, "SELECT value FROM service_state WHERE key=?", key).Scan(&value)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &value, nil
}
func scanSignal(row rowScanner) (DreamSignal, error) {
	var item DreamSignal
	var refs string
	if err := row.Scan(&item.SignalID, &item.SignalType, &refs, &item.Severity, &item.RepoRevision, &item.DedupeKey, &item.Status, &item.EvidenceHash, &item.EvidenceJSON, &item.FirstDetectedAt, &item.LastDetectedAt, &item.ResolvedRevision); err != nil {
		return item, err
	}
	if err := json.Unmarshal([]byte(refs), &item.EntityRefs); err != nil {
		return item, err
	}
	return item, nil
}
